### PR TITLE
Add Batch Norm (+ backward) batching rules 

### PR DESCRIPTION
**Pull request by samdow**
Adds batching rules for batch_norm
- [x] Forwards batching (vmap) works
- [x] Backwards batching (vmapvjp & vjpvmap) works

### CODE REVIEW

File: functorch/csrc/BatchRulesNorm.cpp [hunk 1]
```diff
@@ -396,8 +393,72 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> native_layer_norm_backward_plumbing
...
+    const c10::optional<Tensor>& running_mean_opt, optional<int64_t> running_mean_bdim,
+    const c10::optional<Tensor>& running_var_opt, optional<int64_t> running_var_bdim,
+    bool training, double momentum, double eps) {
+    return batch_norm_batch_rule<F, Func>(
+        input, input_bdim, weight_opt, weight_bdim, bias_opt, bias_bdim, 
+        running_mean_opt, running_mean_bdim, running_var_opt, running_var_bdim, training, momentum, eps);
+  }
+};
+
+#define NATIVE_BATCH_NORM_BATCH_RULE(fn) SINGLE_ARG(\
```
Comment (samdow) [inline functorch/csrc/BatchRulesNorm.cpp:443]: If there's a more efficient way to write these, definitely let me know! My templates knowledge is rather limited so I'm just pattern matching against other helpers

File: functorch/csrc/BatchRulesNorm.cpp [hunk 2]
```diff
@@ -396,8 +393,72 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> native_layer_norm_backward_plumbing
...
+template <typename F, F Func>
+struct CudnnBatchNormBatchRuleHelper {
+  static std::tuple<Tensor,optional<int64_t>,Tensor,optional<int64_t>,Tensor,optional<int64_t>,Tensor,optional<int64_t>> apply(
+    const Tensor& input, optional<int64_t> input_bdim,
+    const Tensor& weight_opt, optional<int64_t> weight_bdim,
+    const c10::optional<Tensor>& bias_opt, optional<int64_t> bias_bdim,
+    const c10::optional<Tensor>& running_mean_opt, optional<int64_t> running_mean_bdim,
+    const c10::optional<Tensor>& running_var_opt, optional<int64_t> running_var_bdim,
+    bool training, double momentum, double eps) {
+    auto reserve = at::tensor({0}, input.options().dtype(kByte));
```
Comment (samdow) [inline functorch/csrc/BatchRulesNorm.cpp:420]: Add comment about what we saw in cudnn that suggested this was okay

File: functorch/csrc/BatchRulesNorm.cpp [hunk 3]
```diff
@@ -5,107 +5,121 @@
...
+  c10::MaybeOwned<Tensor> running_mean_maybe_owned = at::borrow_from_optional_tensor(running_mean_opt);
+  auto running_mean_ = *running_mean_maybe_owned;
+  c10::MaybeOwned<Tensor> running_var_maybe_owned = at::borrow_from_optional_tensor(running_var_opt);
+  auto running_var_ = *running_var_maybe_owned;
 
-  optional<Tensor> weight_value;
-  optional<int64_t> weight_bdim;
-  if (weight) {
-    std::tie(weight_value, weight_bdim) = unwrapTensorAtLevel(*weight, cur_level);
+  if (input_bdim.has_value() && ((running_mean_.defined() && !running_mean_bdim.has_value()) || (running_var_.defined() && !running_var_bdim.has_value()))) {
```
Comment (samdow) [inline functorch/csrc/BatchRulesNorm.cpp:69]: Don't need has_values...from a debugging time when I actually just needed more parentheses

File: functorch/csrc/BatchRulesNorm.cpp [hunk 4]
```diff
@@ -5,107 +5,121 @@
...
-    c10::impl::ExcludeDispatchKeyGuard guard(kBatchedKey);
-    auto results = batch_norm_batch_rule(
-        input_value, input_bdim,
-        weight_value, weight_bdim,
-        bias_value, bias_bdim,
-        running_mean_value, running_mean_bdim,
-        running_var_value, running_var_bdim,
-        training, momentum, eps, cudnn_enabled);
-    return makeBatched(std::get<0>(results), std::get<1>(results), cur_level);
+  const auto input_logical_rank = rankWithoutBatchDim(input, input_bdim);
```
Comment (samdow) [inline functorch/csrc/BatchRulesNorm.cpp:92]: input isn't updated...move this down to where it's used

**Review by zou3519 — APPROVED** [review]
This LGTM. There are a few items we should handle as a follow-up:
- nicer way to specify to get_batched_inputs when to not batch a certain input. Maybe this could be a bitmask or something. Ultimately this information should end up as a part of the OpInfo
- performance testing. All the dummy inputs and the reshapes make me nervous -- it's possible that a simple decomposition (like https://gist.github.com/zou3519/44f7a9c92ccbe8c690e19311f4564182, but idk if it's right) may be faster

File: functorch/csrc/BatchRulesNorm.cpp [hunk 5]
```diff
@@ -5,107 +5,292 @@
...
+  if (running_mean_bdim) {
+    running_mean_ = reshape_dim_into(*running_mean_bdim, 0, running_mean_);
+  }
+  if (running_var_bdim) {
+    running_var_ = reshape_dim_into(*running_var_bdim, 0, running_var_);
+  }
+
+  const auto dummy_weight = at::ones(input_.size(1), input_.options());
+  const auto dummy_bias = at::zeros(input_.size(1), input_.options()); // can't use efficientzeros because cudnn expects contiguous
+  const auto result = Func(input_, dummy_weight, dummy_bias, running_mean_, running_var_, training, momentum, eps);
```
[the remark this answers is not in the document]
Comment (samdow) [reply functorch/csrc/BatchRulesNorm.cpp:94]: cudnn and miopen both require a weight tensor by the signature and they'll error that they need strides if you give them an empty tensor

It looks like they also require bias because I get the same strides error message if I don't pass it in so I updated the comment to reflect that

[the remark this answers is not in the document]
Comment (zou3519) [reply functorch/csrc/BatchRulesNorm.cpp:94]: What happens if we call F.batch_norm without a weight and bias using a CUDA tensor input? Does it go the native_batch_norm route or does it hit cudnn?

[the remark this answers is not in the document]
Comment (samdow) [reply functorch/csrc/BatchRulesNorm.cpp:94]: It will route to native_batch_norm (there's a cuda implementation of native batch norm that I haven't looked at much)

File: functorch/csrc/BatchRulesNorm.cpp [hunk 6]
```diff
@@ -5,107 +5,292 @@
...
+    throw std::runtime_error("Batch norm got a batched tensor as input while the running_mean or running_var, which will be updated in place, were not batched.");
+  }
+  auto input_ = input;
+  c10::optional<int64_t> bdim_size;
+  if (input_bdim) {
+    bdim_size = input.size(*input_bdim);
+    input_ = reshape_dim_into(*input_bdim, /*channels dim*/1, input);
+  } else if (running_mean_bdim) {
+    bdim_size = running_mean_.size(*running_mean_bdim);
+    input_ = repeatForUpdatedValues(input_, *bdim_size);
```
[the remark this answers is not in the document]
Comment (samdow) [reply functorch/csrc/BatchRulesNorm.cpp:79]: Yes this is the output shape (repeat will basically give a tensor with a shape that multiplies each dimension by the shape given, like `[input.size(i) * repeat_size[i] for i in range(len(input.shape))]`).

I saw after writing this while I was writing the backwards rule your `ensure_has_bdim` which we can use and then reshape instead but we might not gain a lot since cudnn and miopen still need the input to be contiguous so we might have to make the inputs contiguous again...I'll look at this

File: functorch/csrc/BatchRulesNorm.cpp [hunk 7]
```diff
@@ -5,107 +5,292 @@
...
-  return reshape_dim_into(*tensor_bdim, 0, *tensor);
+  return tensor.view(new_sizes);
 }
 
-std::tuple<at::Tensor,optional<int64_t>>
+static Tensor repeatForUpdatedValues(const Tensor& input, int64_t batch_size) {
+  VmapDimVector repeat_sizes(input.dim(), 1);
+  repeat_sizes[1] = batch_size;
+  return input.repeat(repeat_sizes);
+}
```
[the remark this answers is not in the document]
Comment (samdow) [reply functorch/csrc/BatchRulesNorm.cpp]: Changed to expand + reshape--didn't end up needing contiguous call like I worried we might
